plutus-scb/test/Plutus/SCB/Effects/MultiAgent.hs

```diff
@@ -0,0 +1,194 @@
+{-# LANGUAGE LambdaCase       #-}
+{-# LANGUAGE RankNTypes       #-}
+{-# LANGUAGE TemplateHaskell  #-}
+{-# LANGUAGE TypeOperators    #-}
+module Plutus.SCB.Effects.MultiAgent(
+    -- $multiagent
+    -- * Agent state
+    AgentState
+    , walletState
+    , nodeClientState
+    , chainIndexState
+    , signingProcessState
+    , emptyAgentState
+    -- * Multi-agent effect
+    , MultiAgentSCBEffect(..)
+    , SCBClientEffects
+    , SCBControlEffects
+    , agentAction
+    , agentControlAction
+    , handleMultiAgent
+    ) where
+
+import           Control.Lens
+import           Control.Lens.TH
+import           Control.Monad.Freer
+import           Control.Monad.Freer.Error
+import           Control.Monad.Freer.Extra.Log
+import           Control.Monad.Freer.Extras
+import qualified Control.Monad.Freer.Log         as Log
+import           Control.Monad.Freer.State       (State)
+import           Control.Monad.Freer.TH
+import           Control.Monad.Freer.Writer
+import           Data.Map                        (Map)
+import           Eventful.Store.Memory           (EventMap)
+
+import qualified Cardano.ChainIndex.Types        as CI
+import           Cardano.Node.Follower           (NodeFollowerEffect)
+import qualified Cardano.Node.Follower           as NF
+import qualified Cardano.Node.Mock               as NodeServer
+import qualified Cardano.Node.Types              as NF
+import qualified Cardano.Node.Types              as NodeServer
+
+import           Plutus.SCB.Effects.Contract     (ContractEffect (..))
+import           Plutus.SCB.Effects.ContractTest (handleContractTest)
+import           Plutus.SCB.Effects.EventLog     (EventLogEffect)
+import           Plutus.SCB.Effects.UUID         (UUIDEffect)
+import           Plutus.SCB.Events               (ChainEvent)
+import           Plutus.SCB.Types                (SCBError (..))
+
+import           Wallet.Effects                  (ChainIndexEffect, NodeClientEffect, SigningProcessEffect,
+                                                  WalletEffect)
+import qualified Wallet.Emulator.Chain           as Chain
+import           Wallet.Emulator.ChainIndex      (ChainIndexControlEffect)
+import qualified Wallet.Emulator.ChainIndex      as ChainIndex
+import           Wallet.Emulator.Error           (WalletAPIError)
+import           Wallet.Emulator.MultiAgent      (EmulatorEvent, chainIndexEvent, walletClientEvent, walletEvent)
+import           Wallet.Emulator.NodeClient      (NodeControlEffect)
+import qualified Wallet.Emulator.NodeClient      as NC
+import           Wallet.Emulator.SigningProcess  (SigningProcessControlEffect)
+import qualified Wallet.Emulator.SigningProcess  as SP
+import           Wallet.Emulator.Wallet          (Wallet, WalletState)
+import qualified Wallet.Emulator.Wallet          as Wallet
+
+-- $multiagent
+-- An SCB version of 'Wallet.Emulator.MultiAgent', with agent-specific states and actions on them.
+-- Agents are represented by the 'Wallet' type.
```

Comment:
An agent corresponds to an SCB, right?

Reply:
Yeah good point. I'll add that to the comment.